src/api/mod.rs

```diff
@@ -95,16 +103,27 @@ impl Api {
                         .await?;
 
                     let mut sub = sub;
-                    while let Some(val) = sub.next().await {
-                        if let Ok(val) = val {
-                            let number = get_number(&val)?;
-
-                            let res = client
-                                .request("chain_getBlockHash", vec![number.into()])
-                                .await?;
 
-                            tracing::debug!("New head: {number} {res}");
-                            head_tx.send_replace(Some((res, number)));
+                    loop {
+                        tokio::select! {
+                            Some(Ok(val)) = sub.next() => {
+                                heartbeat.reset();
+
+                                let number = get_number(&val)?;
+
+                                let res = client
+                                    .request("chain_getBlockHash", vec![number.into()])
+                                    .await?;
+
+                                tracing::debug!("New head: {number} {res}");
+                                head_tx.send_replace(Some((res, number)));
+                            }
+                            _ = heartbeat.tick() => {
```

Comment:
@xlc first tick resolves immediately so it needs to happen outside this loop after heartbeat is created

Reply:
yeah I already did that change and haven't push it yet